spark2/src/test/java/org/apache/iceberg/spark/source/TestSelect.java

```diff
@@ -0,0 +1,230 @@
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.types.Types;
+import org.apache.spark.sql.Dataset;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.SparkSession;
+import org.junit.AfterClass;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+
+import static org.apache.iceberg.types.Types.NestedField.optional;
+
+public class TestSelect {
+  private static final HadoopTables TABLES = new HadoopTables(new Configuration());
+  private static final Schema SCHEMA = new Schema(
+      optional(1, "id", Types.IntegerType.get()),
+      optional(2, "data", Types.StringType.get()),
+      optional(3, "doubleVal", Types.DoubleType.get())
+  );
+
+  private static SparkSession spark;
+
+  private static int scanEventCount = 0;
+  private static ScanEvent lastScanEvent = null;
+
+  private Table table;
+
+  static {
+    Listeners.register(event -> {
+      scanEventCount += 1;
+      lastScanEvent = event;
+    }, ScanEvent.class);
+  }
+
+  @BeforeClass
+  public static void startSpark() {
+    spark = SparkSession.builder()
+        .master("local[2]")
+        .getOrCreate();
+  }
+
+  @AfterClass
+  public static void stopSpark() {
+    SparkSession currentSpark = spark;
+    spark = null;
+    currentSpark.stop();
+  }
+
+  @Rule
+  public TemporaryFolder temp = new TemporaryFolder();
+
+  private String tableLocation = null;
+
+  @Before
+  public void init() throws Exception {
+    File tableDir = temp.newFolder();
+    this.tableLocation = tableDir.toURI().toString();
+
+    table = TABLES.create(SCHEMA, tableLocation);
+
+    List<Record> rows = Lists.newArrayList(
+        new Record(1, "a", 1.0),
+        new Record(2, "b", 2.0),
+        new Record(3, "c", Double.NaN)
+    );
+
+    Dataset<Row> df = spark.createDataFrame(rows, Record.class);
+
+    df.select("id", "data", "doubleVal").write()
+        .format("iceberg")
+        .mode("append")
+        .save(tableLocation);
+
+    table.refresh();
+
+    Dataset<Row> results = spark.read()
+        .format("iceberg")
+        .load(tableLocation);
+    results.createOrReplaceTempView("table");
+
+    scanEventCount = 0;
+    lastScanEvent = null;
+  }
+
+  @Test
+  public void testSelect() {
+    List<Record> expected = ImmutableList.of(
+        new Record(1, "a", 1.0), new Record(2, "b", 2.0), new Record(3, "c", Double.NaN));
+
+    Assert.assertEquals("Should return all expected rows", expected, sql("select * from table"));
+  }
+
+  @Test
+  public void testSelectRewrite() {
+    List<Record> expected = ImmutableList.of(new Record(3, "c", Double.NaN));
+
+    Assert.assertEquals("Should return all expected rows", expected,
+        sql("SELECT * FROM table where doubleVal = double('NaN')"));
+    Assert.assertEquals("Should create only one scan", 1, scanEventCount);
```

Comment:
Shouldn't this validate more than just the number of scans?

Reply:
Yes, sorry I forgot to revisit this after cleaning up other changes. Since in spark2 we don't have `Spark3Util.describe()` I wasn't sure to which level we want to assert the expression, so that we can still have test coverage without being too coupled with internal implementation. Let me know how you think the updated test is!

Reply:
Looks good!